File: parser.py
import csv
import html as html_mod
import io
from datetime import datetime
from categorizer import categorize

BANK_FORMATS = {
    "capital_one_credit": "Capital One Credit Card",
    "capital_one_bank":   "Capital One Checking / Savings",
    "uccu_checking":      "UCCU Checking",
}


def _parse_date(raw: str) -> str:
    for fmt in ("%Y-%m-%d", "%m/%d/%Y", "%m/%d/%y", "%m-%d-%Y", "%Y/%m/%d"):
        try:
            return datetime.strptime(raw.strip(), fmt).strftime("%Y-%m-%d")
        except ValueError:
            continue
    return raw.strip()


def _normalize_amount(raw: str) -> float:
    return abs(float(raw.replace(",", "").replace("$", "").strip() or 0))


# Prefixes Capital One bank CSVs prepend before the actual merchant name
_CARD_PREFIXES = [
    "Debit Card Purchase - ",
    "Digital Card Purchase - ",
    "360 Checking Card Adjustment Signature (Credit) ",
    "360 Checking Card Adjustment Signature (Debit) ",
]

# Transaction type strings that legitimately contain " - " (not user notes)
_KNOWN_TX_TYPES = [
    "debit card purchase",
    "digital card purchase",
    "360 checking card adjustment",
]


def _clean_bank_description(raw: str) -> tuple[str, str]:
    """
    Returns (cleaned_description, user_note).

    Capital One lets users add notes to transfers in the format:
      "user note here - Deposit from Account XXXXXXX1234"
    This splits those out so the note lands in the notes field.
    """
    note = ""
    desc = raw.strip()

    # Extract user note: anything before " - " that isn't a known transaction type prefix
    if " - " in desc:
        before, after = desc.split(" - ", 1)
        if not any(before.lower().startswith(t) for t in _KNOWN_TX_TYPES):
            note = before.strip()
            desc = after.strip()

    # Strip card purchase type prefixes to get the raw merchant name
    for prefix in _CARD_PREFIXES:
        if desc.lower().startswith(prefix.lower()):
            desc = desc[len(prefix):]
            break

    return desc.strip(), note


def _is_internal_transfer(description: str) -> bool:
    """
    Capital One masks account numbers as XXXXXXX#### in transfer descriptions.
    Any transaction referencing a masked account is an internal transfer.
    """
    return "xxxxxxx" in description.lower()


def detect_and_parse(file_bytes: bytes, bank_format: str, merchant_rules: list = None) -> list[dict]:
    text = file_bytes.decode("utf-8-sig", errors="replace")
    reader = csv.DictReader(io.StringIO(text))

    if bank_format == "capital_one_credit":
        return _parse_credit(reader, merchant_rules)
    elif bank_format == "capital_one_bank":
        return _parse_bank(reader, merchant_rules)
    elif bank_format == "uccu_checking":
        return _parse_uccu(reader, merchant_rules)
    else:
        raise ValueError(f"Unknown bank format: {bank_format}")


def _parse_credit(reader, merchant_rules=None) -> list[dict]:
    """
    Capital One Savor credit card CSV:
      Transaction Date, Posted Date, Card No., Description, Category, Debit, Credit
    """
    rows = []
    for raw in reader:
        row = {k.strip().lower(): (v or "").strip() for k, v in raw.items()}

        date_raw = row.get("transaction date") or row.get("date") or ""
        description = row.get("description") or row.get("merchant name") or ""
        debit = row.get("debit") or ""
        credit = row.get("credit") or ""
        raw_cat = row.get("category") or ""

        if not date_raw or not description:
            continue

        if debit:
            amount = _normalize_amount(debit)
            flow = "debit"
        elif credit:
            amount = _normalize_amount(credit)
            flow = "credit"
        else:
            continue

        rows.append({
            "date": _parse_date(date_raw),
            "description": description,
            "amount": round(amount, 2),
            "flow": flow,
            "raw_category": raw_cat,
            "category": categorize(description, raw_cat, merchant_rules),
            "notes": "",
        })
    return rows


def _parse_bank(reader, merchant_rules=None) -> list[dict]:
    """
    Capital One 360 Checking / Savings CSV:
      Account Number, Transaction Description, Transaction Date,
      Transaction Type, Transaction Amount, Balance
    """
    rows = []
    for raw in reader:
        row = {k.strip().lower(): (v or "").strip() for k, v in raw.items()}

        date_raw = (
            row.get("transaction date")
            or row.get("date")
            or row.get("posted date")
            or ""
        )
        raw_description = (
            row.get("transaction description")
            or row.get("description")
            or row.get("memo")
            or ""
        )
        amount_raw = row.get("transaction amount") or row.get("amount") or ""
        tx_type = row.get("transaction type") or ""

        if not date_raw or not raw_description or not amount_raw:
            continue

        try:
            amount_val = float(amount_raw.replace(",", "").replace("$", "").strip())
        except ValueError:
            continue

        # Determine flow from Transaction Type column (most reliable for Cap One bank CSVs)
        if tx_type.lower() in ("debit", "withdrawal", "purchase"):
            flow = "debit"
        elif tx_type.lower() in ("credit", "deposit"):
            flow = "credit"
        else:
            flow = "debit" if amount_val < 0 else "credit"

        # Clean description and extract any user note
        description, note = _clean_bank_description(raw_description)

        # Internal transfers (masked account numbers) always categorize as Transfer
        if _is_internal_transfer(raw_description):
            category = "Transfer"
        else:
            category = categorize(description, merchant_rules=merchant_rules)

        rows.append({
            "date": _parse_date(date_raw),
            "description": description,
            "amount": round(abs(amount_val), 2),
            "flow": flow,
            "raw_category": "",
            "category": category,
            "notes": note,
        })
    return rows


def _parse_uccu(reader, merchant_rules=None) -> list[dict]:
    """
    UCCU Checking CSV:
      Account Number, Post Date, Check, Description, Debit, Credit, Status, Balance, Classification
    - Dates: M/D/YYYY
    - Separate Debit / Credit columns (one is empty per row)
    - Classification: UCCU's own category label (may contain HTML entities like &amp;)
    - Only import rows with Status == 'Posted'
    """
    rows = []
    for raw in reader:
        row = {k.strip().lower(): (v or "").strip() for k, v in raw.items()}

        date_raw       = row.get("post date") or ""
        description    = row.get("description") or ""
        debit          = row.get("debit") or ""
        credit         = row.get("credit") or ""
        status         = row.get("status") or ""
        classification = html_mod.unescape(row.get("classification") or "")

        # Skip pending / unposted rows
        if status.lower() != "posted":
            continue

        if not date_raw or not description:
            continue

        if debit:
            amount = _normalize_amount(debit)
            flow = "debit"
        elif credit:
            amount = _normalize_amount(credit)
            flow = "credit"
        else:
            continue

        # UCCU explicitly labels transfers — trust it
        if classification.lower() == "transfer":
            category = "Transfer"
        else:
            category = categorize(description, classification, merchant_rules)

        rows.append({
            "date": _parse_date(date_raw),
            "description": description,
            "amount": round(amount, 2),
            "flow": flow,
            "raw_category": classification,
            "category": category,
            "notes": "",
        })
    return rows

File: web/lib/parser.ts
/**
 * Multi-bank CSV parsing — a faithful TypeScript port of the legacy `parser.py`.
 *
 * `detectAndParse(input, bankFormat, merchantRules)` routes to the right
 * sub-parser based on the account's bank format and returns normalized rows.
 * Pure module: only depends on papaparse + the pure categorizer. No DB/framework.
 *
 * Invariants (see CLAUDE.md): amounts are always positive; direction lives in
 * `flow`; internal transfers categorize as "Transfer".
 */

import Papa from "papaparse";
import { categorize, type MerchantRule } from "./categorizer";

export type Flow = "debit" | "credit";

export interface ParsedRow {
  date: string;
  description: string;
  amount: number;
  flow: Flow;
  raw_category: string;
  category: string;
  notes: string;
}

export const BANK_FORMATS = {
  capital_one_credit: "Capital One Credit Card",
  capital_one_bank: "Capital One Checking / Savings",
  uccu_checking: "UCCU Checking",
} as const;

export type BankFormat = keyof typeof BANK_FORMATS;

// Prefixes Capital One bank CSVs prepend before the actual merchant name
const CARD_PREFIXES = [
  "Debit Card Purchase - ",
  "Digital Card Purchase - ",
  "360 Checking Card Adjustment Signature (Credit) ",
  "360 Checking Card Adjustment Signature (Debit) ",
];

// Transaction type strings that legitimately contain " - " (not user notes)
const KNOWN_TX_TYPES = [
  "debit card purchase",
  "digital card purchase",
  "360 checking card adjustment",
];

function round2(n: number): number {
  return Math.round((n + Number.EPSILON) * 100) / 100;
}

/** Parse a date in any of the known bank formats to YYYY-MM-DD, else return it trimmed. */
export function parseDate(raw: string): string {
  const s = (raw ?? "").trim();
  const pad = (v: string) => v.padStart(2, "0");
  // YYYY-MM-DD or YYYY/MM/DD
  let m = s.match(/^(\d{4})[/-](\d{1,2})[/-](\d{1,2})$/);
  if (m) return `${m[1]}-${pad(m[2])}-${pad(m[3])}`;
  // M/D/YYYY or M-D-YYYY
  m = s.match(/^(\d{1,2})[/-](\d{1,2})[/-](\d{4})$/);
  if (m) return `${m[3]}-${pad(m[1])}-${pad(m[2])}`;
  // M/D/YY -> 20YY (bank exports are recent)
  m = s.match(/^(\d{1,2})[/-](\d{1,2})[/-](\d{2})$/);
  if (m) return `20${m[3]}-${pad(m[1])}-${pad(m[2])}`;
  return s;
}

function normalizeAmount(raw: string): number {
  const cleaned = (raw ?? "").replace(/,/g, "").replace(/\$/g, "").trim();
  const n = parseFloat(cleaned);
  return Math.abs(Number.isNaN(n) ? 0 : n);
}

/** Minimal HTML entity unescape for UCCU classification labels (e.g. `&amp;`). */
function htmlUnescape(s: string): string {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

/**
 * Returns [cleanedDescription, userNote]. Capital One lets users add notes to
 * transfers as "user note - Deposit from Account XXXXXXX1234"; this splits them.
 */
export function cleanBankDescription(raw: string): [string, string] {
  let note = "";
  let desc = (raw ?? "").trim();

  if (desc.includes(" - ")) {
    const idx = desc.indexOf(" - ");
    const before = desc.slice(0, idx);
    const after = desc.slice(idx + 3);
    if (!KNOWN_TX_TYPES.some((t) => before.toLowerCase().startsWith(t))) {
      note = before.trim();
      desc = after.trim();
    }
  }

  for (const prefix of CARD_PREFIXES) {
    if (desc.toLowerCase().startsWith(prefix.toLowerCase())) {
      desc = desc.slice(prefix.length);
      break;
    }
  }

  return [desc.trim(), note];
}

/** Capital One masks account numbers as XXXXXXX#### in transfer descriptions. */
function isInternalTransfer(description: string): boolean {
  return description.toLowerCase().includes("xxxxxxx");
}

type Row = Record<string, string>;

/** Lowercase + trim every key, coerce values to trimmed strings. */
function normRow(raw: Record<string, unknown>): Row {
  const out: Row = {};
  for (const [k, v] of Object.entries(raw)) {
    out[k.trim().toLowerCase()] = (v == null ? "" : String(v)).trim();
  }
  return out;
}

function parseCsv(input: string | Uint8Array): Row[] {
  const text = (typeof input === "string" ? input : new TextDecoder("utf-8").decode(input))
    .replace(/^﻿/, ""); // strip BOM (utf-8-sig)
  const result = Papa.parse<Record<string, unknown>>(text, {
    header: true,
    skipEmptyLines: true,
  });
  return result.data.map(normRow);
}

export function detectAndParse(
  input: string | Uint8Array,
  bankFormat: BankFormat | string,
  merchantRules: MerchantRule[] | null = null,
): ParsedRow[] {
  const rows = parseCsv(input);
  switch (bankFormat) {
    case "capital_one_credit":
      return parseCredit(rows, merchantRules);
    case "capital_one_bank":
      return parseBank(rows, merchantRules);
    case "uccu_checking":
      return parseUccu(rows, merchantRules);
    default:
      throw new Error(`Unknown bank format: ${bankFormat}`);
  }
}

/** Capital One Savor credit card: Transaction Date, Posted Date, Card No., Description, Category, Debit, Credit */
function parseCredit(rows: Row[], merchantRules: MerchantRule[] | null): ParsedRow[] {
  const out: ParsedRow[] = [];
  for (const row of rows) {
    const dateRaw = row["transaction date"] || row["date"] || "";
    const description = row["description"] || row["merchant name"] || "";
    const debit = row["debit"] || "";
    const credit = row["credit"] || "";
    const rawCat = row["category"] || "";

    if (!dateRaw || !description) continue;

    let amount: number;
    let flow: Flow;
    if (debit) {
      amount = normalizeAmount(debit);
      flow = "debit";
    } else if (credit) {
      amount = normalizeAmount(credit);
      flow = "credit";
    } else {
      continue;
    }

    out.push({
      date: parseDate(dateRaw),
      description,
      amount: round2(amount),
      flow,
      raw_category: rawCat,
      category: categorize(description, rawCat, merchantRules),
      notes: "",
    });
  }
  return out;
}

/** Capital One 360 Checking/Savings: Account Number, Transaction Description, Transaction Date, Transaction Type, Transaction Amount, Balance */
function parseBank(rows: Row[], merchantRules: MerchantRule[] | null): ParsedRow[] {
  const out: ParsedRow[] = [];
  for (const row of rows) {
    const dateRaw =
      row["transaction date"] || row["date"] || row["posted date"] || "";
    const rawDescription =
      row["transaction description"] || row["description"] || row["memo"] || "";
    const amountRaw = row["transaction amount"] || row["amount"] || "";
    const txType = row["transaction type"] || "";

    if (!dateRaw || !rawDescription || !amountRaw) continue;

    const amountVal = parseFloat(amountRaw.replace(/,/g, "").replace(/\$/g, "").trim());
    if (Number.isNaN(amountVal)) continue;

    let flow: Flow;
    const t = txType.toLowerCase();
    if (t === "debit" || t === "withdrawal" || t === "purchase") {
      flow = "debit";
    } else if (t === "credit" || t === "deposit") {
      flow = "credit";
    } else {
      flow = amountVal < 0 ? "debit" : "credit";
    }

    const [description, note] = cleanBankDescription(rawDescription);
    const category = isInternalTransfer(rawDescription)
      ? "Transfer"
      : categorize(description, "", merchantRules);

    out.push({
      date: parseDate(dateRaw),
      description,
      amount: round2(Math.abs(amountVal)),
      flow,
      raw_category: "",
      category,
      notes: note,
    });
  }
  return out;
}

/** UCCU Checking: Account Number, Post Date, Check, Description, Debit, Credit, Status, Balance, Classification */
function parseUccu(rows: Row[], merchantRules: MerchantRule[] | null): ParsedRow[] {
  const out: ParsedRow[] = [];
  for (const row of rows) {
    const dateRaw = row["post date"] || "";
    const description = row["description"] || "";
    const debit = row["debit"] || "";
    const credit = row["credit"] || "";
    const status = row["status"] || "";
    const classification = htmlUnescape(row["classification"] || "");

    if (status.toLowerCase() !== "posted") continue;
    if (!dateRaw || !description) continue;

    let amount: number;
    let flow: Flow;
    if (debit) {
      amount = normalizeAmount(debit);
      flow = "debit";
    } else if (credit) {
      amount = normalizeAmount(credit);
      flow = "credit";
    } else {
      continue;
    }

    const category =
      classification.toLowerCase() === "transfer"
        ? "Transfer"
        : categorize(description, classification, merchantRules);

    out.push({
      date: parseDate(dateRaw),
      description,
      amount: round2(amount),
      flow,
      raw_category: classification,
      category,
      notes: "",
    });
  }
  return out;
}
